### PR TITLE
fix: correct the range behavior in MemoryKvBackend & RaftEngineBackend

### LINKED ISSUE #2616: Correct the behaviors in KvBackend implements
### What type of enhancement is this?

Tech debt reduction

### What does the enhancement do?

The `MemoryKvBackend` and `RaftEngineBackend` don't correctly handle the `range` & `delete_range`. 

The KvBackend trait requires: 
- If range_end is '\0', the range is all keys >= key.
- If both key and range_end are '\0', then the range request returns all.

### Steps
- [ ] https://github.com/GreptimeTeam/greptimedb/pull/2615
- [ ] #2618

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Correct the range behavior in MemoryKvBackend & RaftEngineBackend

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
close #2616

### CODE REVIEW

File: src/common/meta/src/kv_backend/memory.rs [hunk 1]
```diff
@@ -96,6 +96,14 @@ impl<T: ErrorExt + Send + Sync + 'static> KvBackend for MemoryKvBackend<T> {
...
         let iter: Box<dyn Iterator<Item = (&Vec<u8>, &Vec<u8>)>> = if range_end.is_empty() {
             Box::new(kvs.get_key_value(&key).into_iter())
+        } else if range_end.len() == 1 && range_end[0] == 0 {
+            // If both key and range_end are ‘\0’, then range represents all keys.
+            if key.len() == 1 && key[0] == 0 {
+                Box::new(kvs.iter())
+            } else {
+                // If range_end is ‘\0’, the range is all keys greater than or equal to the key argument.
+                Box::new(kvs.range(key..))
+            }
```
Comment (fengjiachun) [inline src/common/meta/src/kv_backend/memory.rs:106]: Does the `RaftEngineBackend` also need to be synchronized?

Comment (WenyXu) [reply src/common/meta/src/kv_backend/memory.rs:106]: Yes, and we also need to fix the `delete_range`.

File: src/common/meta/src/rpc/store.rs [hunk 2]
```diff
@@ -30,6 +31,22 @@ use crate::error;
...
+        // If both key and range_end are ‘\0’, then range represents all keys.
+        if key.len() == 1 && key[0] == 0 {
+            (Bound::Unbounded, Bound::Unbounded)
+        } else {
+            // If range_end is ‘\0’, the range is all keys greater than or equal to the key argument.
+            (Bound::Included(key), Bound::Unbounded)
+        }
+    } else {
+        (Bound::Included(key), Bound::Excluded(range_end))
+    }
```
Comment (MichaelScofield) [inline src/common/meta/src/rpc/store.rs:47]:
```rust
match (&key[..], &range_end[..]) {
  (_, []) => (Bound::Included(key.clone()), Bound::Included(key)),
  ([0], [0]) => (Bound::Unbounded, Bound::Unbounded),
  (_, [0]) => (Bound::Included(key), Bound::Unbounded),
  (_, _) => (Bound::Included(key), Bound::Excluded(range_end)),
}
```

File: src/common/meta/src/kv_backend/memory.rs [hunk 3]
```diff
@@ -215,31 +208,23 @@ impl<T: ErrorExt + Send + Sync + 'static> KvBackend for MemoryKvBackend<T> {
...
-                end: range_end,
-            };
-            kvs.extract_if(|key, _| range.contains(key))
-                .map(Into::into)
-                .collect::<Vec<_>>()
-        };
+        let keys = kvs
+            .range(range)
+            .map(|(key, _)| key.clone())
+            .collect::<Vec<_>>();
```
Comment (MichaelScofield) [inline src/common/meta/src/kv_backend/memory.rs:219]: Collect costs memory. Is it possible to use the old "extract_if"?

Comment (WenyXu) [reply src/common/meta/src/kv_backend/memory.rs:219]: > Collect costs memory. Is it possible to use the old "extract_if"?

But the `extract_if ` needs to visit all elements.

File: src/common/meta/src/rpc/store.rs (hunk 2 above)
Comment (WenyXu) [reply src/common/meta/src/rpc/store.rs]: It will panic if the key is empty.

**Comment by WenyXu** [comment]
@MichaelScofield @fengjiachun PTAL

File: src/common/meta/src/kv_backend/memory.rs [hunk 4]
```diff
@@ -85,35 +84,32 @@ impl<T: ErrorExt + Send + Sync + 'static> KvBackend for MemoryKvBackend<T> {
...
+        let iter = kvs.range(range);
 
-        let iter: Box<dyn Iterator<Item = (&Vec<u8>, &Vec<u8>)>> = if range_end.is_empty() {
-            Box::new(kvs.get_key_value(&key).into_iter())
-        } else {
-            Box::new(kvs.range(key..range_end))
-        };
-        let mut kvs = iter
+        let mut more = false;
+        let mut counter = 0usize;
```
Comment (fengjiachun) [inline src/common/meta/src/kv_backend/memory.rs:96]:
```suggestion
        let mut took = 0;
```

File: src/common/meta/src/kv_backend/memory.rs [hunk 5]
```diff
@@ -85,35 +84,32 @@ impl<T: ErrorExt + Send + Sync + 'static> KvBackend for MemoryKvBackend<T> {
...
+
+        let kvs = iter
+            .take_while(|_| {
+                let take = counter != limit as usize || limit == 0;
+                counter += 1;
+                more = counter > limit as usize && limit != 0;
+
+                take
+            })
```
Comment (fengjiachun) [inline src/common/meta/src/kv_backend/memory.rs:105]: suggestion:

```
let take = limit <= 0 || took != limit;
took += 1;
more = limit > 0 && took > limit;

take
```

File: src/log-store/src/raft_engine/backend.rs [hunk 6]
```diff
@@ -137,29 +139,48 @@ impl KvBackend for RaftEngineBackend {
...
     async fn range(&self, req: RangeRequest) -> Result<RangeResponse, Self::Error> {
         let mut res = vec![];
+        let (start, end) = req.range();
+        let RangeRequest {
+            keys_only, limit, ..
+        } = req;
+
+        let (start_key, end_key) = match (start, end) {
+            (Included(start), Included(_)) => {
+                (Some(start.clone()), Some(get_next_prefix_key(&start)))
```
Comment (fengjiachun) [inline src/log-store/src/raft_engine/backend.rs:149]:
```suggestion
                (Some(start), Some(end))
```

Comment (WenyXu) [reply src/log-store/src/raft_engine/backend.rs:149]: The raft engine use the [start,end) pattern.

File: src/common/meta/src/kv_backend/memory.rs [hunk 7]
```diff
@@ -85,35 +84,32 @@ impl<T: ErrorExt + Send + Sync + 'static> KvBackend for MemoryKvBackend<T> {
...
-            Box::new(kvs.get_key_value(&key).into_iter())
-        } else {
-            Box::new(kvs.range(key..range_end))
-        };
-        let mut kvs = iter
+        let mut more = false;
+        let mut counter = 0usize;
+
+        let kvs = iter
+            .take_while(|_| {
```
Comment (MichaelScofield) [inline src/common/meta/src/kv_backend/memory.rs:99]: Can we simply use `take(limit)` (should check if limit > 0 first of course)?

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/meta/src/kv_backend/memory.rs (hunk 3 above)
Comment (fengjiachun) [reply src/common/meta/src/kv_backend/memory.rs:222]: Can we use iter directly without collecting to vec?

File: src/common/meta/src/kv_backend/memory.rs (hunk 7 above)
Comment (fengjiachun) [reply src/common/meta/src/kv_backend/memory.rs:99]: If we use take(), it seems cannot calculate `more`

File: src/common/meta/src/kv_backend/memory.rs [hunk 8]
```diff
@@ -215,32 +211,30 @@ impl<T: ErrorExt + Send + Sync + 'static> KvBackend for MemoryKvBackend<T> {
...
+            if let Some(value) = kvs.remove(&key) {
+                if prev_kv {
+                    prev_kvs.push((key.clone(), value).into())
+                }
+            }
+        }
+
         Ok(DeleteRangeResponse {
             deleted: prev_kvs.len() as i64,
             prev_kvs: if prev_kv { prev_kvs } else { vec![] },
```
Comment (fengjiachun) [inline src/common/meta/src/kv_backend/memory.rs:240]:
```suggestion
            prev_kvs,
```

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**